### PR TITLE
Added support for system node pools in managedClusters

### LINKED ISSUE #1376: [AKS] The default MachinePool should be of "system" type, not "user"
/kind bug

**What steps did you take and what happened:**
I create an AKS cluster with 2 machine pools.

**What did you expect to happen:**
The default machine pool should be a "system" node pool.

**Anything else you would like to add:**
The node pool type should be a property of `AzureManagedMachinePool`, in order to allow its configuration.

**Environment:**
- cluster-api-provider-azure version: `v0.4.14`
- Kubernetes version: (use `kubectl version`): `v1.20.5`
- OS (e.g. from `/etc/os-release`): `Ubuntu 18.04.2 LTS`

### PR DESCRIPTION
**Pull request by LochanRn**
**What type of PR is this?**
/kind bug
/area managedclusters

**What this PR does / why we need it**:
system node pool will be of type system and not user.
added mode in the azuremannagedmachinepool api to set the mode of azure agent pools
removed defaultPoolRef from azuremanagedcontrolplane api. (provides more flexibility to manage system node pools.)
validation for deletion of last system node pool is handled in the webhook.
bumped up the containerservice sdk from 2020-02-01 to 2021-05-01
updated the aks docs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes  #1376 #1418 #1536

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Mode spec in the AzureManagedMachinePool is used to specify the mode of an agentPool i.e System or User.
Removed defaultPoolRef from AzureManagedControlPlane. 
```

### CODE REVIEW

File: test/e2e/config/azure-dev.yaml [hunk 1]
```diff
@@ -118,7 +118,7 @@ variables:
 intervals:
   default/wait-controllers: ["3m", "10s"]
   default/wait-cluster: ["20m", "10s"]
-  default/wait-control-plane: ["20m", "10s"]
+  default/wait-control-plane: ["35m", "10s"]
```
Comment (CecileRobertMichon) [inline test/e2e/config/azure-dev.yaml:121]: can we revert this change?

Comment (LochanRn) [reply test/e2e/config/azure-dev.yaml:121]: ack

File: docs/book/src/topics/managedcluster.md [hunk 2]
```diff
@@ -12,9 +12,9 @@ custom resources:
 The combination of AzureManagedControlPlane/AzureManagedCluster
 corresponds to provisioning an AKS cluster. AzureManagedMachinePool
 corresponds one-to-one with AKS node pools. This also means that
-creating an AzureManagedControlPlane requires defining the default
-machine pool, since AKS requires at least one system pool at creation
-time.
+creating an AzureManagedCluster requires at least one AzureManagedMachinePool
```
Comment (alexeldeib) [inline docs/book/src/topics/managedcluster.md:15]: probably leave this as AzureManagedControlPlane, since I have a PR open to delete AzureManagedCluster entirely

File: docs/book/src/topics/managedcluster.md [hunk 3]
```diff
@@ -95,13 +95,11 @@ kind: AzureManagedControlPlane
...
   name: my-cluster-control-plane
 spec:
-  defaultPoolRef:
-    name: agentpool0
   location: southcentralus
   resourceGroup: foo-bar
   sshPublicKey: ${AZURE_SSH_PUBLIC_KEY_B64:=""}
   subscriptionID: fae7cc14-bfba-4471-9435-f945b42a16dd # fake uuid
-  version: v1.19.6
+  version: v1.20.5
```
Comment (alexeldeib) [inline docs/book/src/topics/managedcluster.md:102]: thank you!

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+// SetupWebhookWithManager sets up and registers the webhook with the manager.
+func (r *AzureManagedMachinePool) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// NewAzureManagedMachinePoolWebhook returns a new AzureManagedMachinePool instance.
+func NewAzureManagedMachinePoolWebhook(client client.Client) *AzureManagedMachinePool {
+	webhookClient = client
```
Comment (alexeldeib) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:50]: please do not make this a package level variable set inside a constructor.

I think you should use the lower level webhook implementation like in https://github.com/alexeldeib/hook-test/blob/master/webhook.go

sorry I missed this on earlier reviews

Comment (alexeldeib) [reply exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:50]: you might be able to get away with wrapping the CRD type with a "fake" type which implements CRD + Client and implements Defaulter, but I'm not sure controller-runtime will be happy with that.

File: exp/controllers/azuremangedcontrolplane_reconciler.go [hunk 5]
```diff
@@ -209,27 +210,63 @@ func (r *azureManagedControlPlaneReconciler) reconcileManagedCluster(ctx context
...
-		if scope.InfraMachinePool.Spec.OSDiskSizeGB != nil {
-			defaultPoolSpec.OSDiskSizeGB = *scope.InfraMachinePool.Spec.OSDiskSizeGB
+		if err := r.kubeclient.List(ctx, ammpList, opt1, opt2); err != nil {
+			return err
 		}
-		if scope.MachinePool.Spec.Replicas != nil {
-			defaultPoolSpec.Replicas = *scope.MachinePool.Spec.Replicas
+
+		if ammpList == nil || len(ammpList.Items) == 0 {
+			return errors.New("failed to fetch azuremanagedMachine pool with mode:System, require atleast 1 system node pool")
```
Comment (alexeldeib) [inline exp/controllers/azuremangedcontrolplane_reconciler.go:230]:
```suggestion
			return errors.New("failed to fetch azuremanagedMachine pool with mode:System, require at least 1 system node pool")
```

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go (hunk 4 above)
Comment (alexeldeib) [reply exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:50]: yeah looks like no dice, CR will check if the object it got the webhook request for implements validator/defaulter: https://github.com/kubernetes-sigs/controller-runtime/blob/8b55f85c90c3b1df1af58ddb7fe50096fdc7aa99/pkg/webhook/admission/validator.go#L64

File: docs/book/src/topics/managedcluster.md (hunk 2 above)
Comment (LochanRn) [reply docs/book/src/topics/managedcluster.md:15]: ack

File: exp/controllers/azuremangedcontrolplane_reconciler.go (hunk 5 above)
Comment (LochanRn) [reply exp/controllers/azuremangedcontrolplane_reconciler.go:230]: ack

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go (hunk 4 above)
Comment (LochanRn) [reply exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:50]: okay sure let me change this.

**Comment by LochanRn** [comment]
@CecileRobertMichon  @alexeldeib the cluster-api-provider-azure-e2e-exp job timesout.
locally when I increase the time for default/wait-control-plane to ["30m", "10s"]. It does not timeout.

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var azuremanagedmachinepoollog = logf.Log.WithName("azuremanagedmachinepool-resource")
+
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-azuremanagedmachinepool,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedmachinepools,verbs=create;update,versions=v1alpha4,name=default.azuremanagedmachinepools.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
```
Comment (alexeldeib) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:37]:
```suggestion
//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-azuremanagedmachinepool,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedmachinepools,verbs=create;update,versions=v1alpha4,name=default.azuremanagedmachinepools.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
```

and regenerate webhook manifests please

Comment (alexeldeib) [reply exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:37]: also matchPolicy equivalent (I don't think webhooks enforce a default)

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+func (r *AzureManagedMachinePool) Default(client client.Client) {
+	azuremanagedmachinepoollog.Info("default", "name", r.Name)
+
+	if r.Labels == nil {
+		r.Labels = make(map[string]string)
+	}
+	r.Labels[LabelAgentPoolMode] = r.Spec.Mode
+}
+
+//+kubebuilder:webhook:verbs=update;delete,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha4-azuremanagedmachinepool,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedmachinepools,versions=v1alpha4,name=validation.azuremanagedmachinepools.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
```
Comment (alexeldeib) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:49]: same as above

File: util/webhook/validator.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	v1 "k8s.io/api/admission/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// Validator defines functions for validating an operation.
+type Validator interface {
```
Comment (alexeldeib) [inline util/webhook/validator.go:35]: very nice 😎

File: util/webhook/validator.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+// Handle handles admission requests.
+func (h *validatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if h.validator == nil {
+		panic("validator should never be nil")
+	}
+
+	// Get the object in the request
+	obj := h.validator.DeepCopyObject().(Validator)
+	if req.Operation == v1.Create {
+		err := h.decoder.Decode(req, obj)
```
Comment (alexeldeib) [inline util/webhook/validator.go:75]: is there a reason to use decode(req, obj) here but decodeRaw(req.Object, obj) below? I see controller-runtime does the same, but decode(req, obj) only does a length check before calling decodeRaw() anyway: https://github.com/kubernetes-sigs/controller-runtime/blob/1e4d87c9f9e15e4a58bb81909dd787f30ede7693/pkg/webhook/admission/decode.go#L42-L48

fine as-is, just was wondering.

File: util/webhook/validator.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"net/http"
+
+	goerrors "errors"
```
Comment (alexeldeib) [inline util/webhook/validator.go:23]: nit: I wouldn't alias this

File: util/webhook/validator.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+			}
+			return admission.Denied(err.Error())
+		}
+	}
+
+	return admission.Allowed("")
+}
+
+func validationResponseFromStatus(allowed bool, status metav1.Status) admission.Response {
+	resp := admission.Response{
```
Comment (alexeldeib) [inline util/webhook/validator.go:134]: nit, can just return the struct directly

**Comment by alexeldeib** [comment]
mostly nits, webhook config is the only real change. happy to sign off after that.

**Comment by alexeldeib** [comment]
and this was a ton of work, thanks so much! great to have this in CAPZ

**Comment by alexeldeib** [comment]
> @CecileRobertMichon @alexeldeib the cluster-api-provider-azure-e2e-exp job timesout.
locally when I increase the time for default/wait-control-plane to ["30m", "10s"]. It does not timeout.

if you really want to bump the interval before I merge the fixes from #1488 as a separate PR, that's fine, but please make it a separate interval which is not re-used by any other tests. We don't want to reduce the bar that we already set for CAPZ machine tests.

File: util/webhook/validator.go (hunk 10 above)
Comment (LochanRn) [reply util/webhook/validator.go:23]: ack

File: util/webhook/validator.go (hunk 11 above)
Comment (LochanRn) [reply util/webhook/validator.go:134]: ack

**Comment by LochanRn** [comment]
> and this was a ton of work, thanks so much! great to have this in CAPZ

Extremely happy to contribute as well 😊

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go (hunk 6 above)
Comment (LochanRn) [reply exp/api/v1alpha4/azuremanagedmachinepool_webhook.go]: ack.

**Comment by CecileRobertMichon** [comment]
@alexeldeib thanks for all the reviews here, let me know when you're ready for me to do a final pass

@LochanRn awesome work

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+		LabelAgentPoolMode:         string(NodePoolModeSystem),
+	})
+
+	ammpList := &AzureManagedMachinePoolList{}
+	if err := cli.List(ctx, ammpList, opt1, opt2); err != nil {
+		return err
+	}
+
+	if len(ammpList.Items) <= 1 {
+		return errors.New("Aks Cluster must have at least one system pool")
```
Comment (CecileRobertMichon) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:157]:
```suggestion
		return errors.New("AKS Cluster must have at least one system pool")
```

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+	if r.Spec.Mode != string(NodePoolModeSystem) {
+		return nil
+	}
+
+	return r.validateLastSystemNodePool(client)
+}
+
+// validateLastSystemNodePool is used to check if the existing system node pool is the last system node pool.
+// If it is a last system node pool it cannot be deleted or mutated to user node pool as aks expects min 1 system node pool.
```
Comment (CecileRobertMichon) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:118]:
```suggestion
// If it is a last system node pool it cannot be deleted or mutated to user node pool as AKS expects min 1 system node pool.
```

File: util/webhook/validator.go (hunk 9 above)
Comment (CecileRobertMichon) [reply util/webhook/validator.go:75]: @LochanRn do you have an answer for the above?

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

as part of the slack discussion I remember we had discussed documenting mitigation steps in case someone gets into the bad state of trying to delete the last system MachinePool and their MP is stuck in Deleting, can we file an issue to do this as part of a follow up if it's not done in this PR?

File: util/webhook/validator.go (hunk 9 above)
Comment (LochanRn) [reply util/webhook/validator.go:75]: @CecileRobertMichon 
I picked this up from the container-runtime and this looks reasonable enough I guess.
https://github.com/kubernetes-sigs/controller-runtime/blob/1e4d87c9f9e15e4a58bb81909dd787f30ede7693/pkg/webhook/admission/decode.go#L39-L42

**Comment by LochanRn** [comment]
> lgtm
> 
> as part of the slack discussion I remember we had discussed documenting mitigation steps in case someone gets into the bad state of trying to delete the last system MachinePool and their MP is stuck in Deleting, can we file an issue to do this as part of a follow up if it's not done in this PR?

Sure I will file an Issue for documenting it.

**Comment by LochanRn** [comment]
> lgtm
> 
> as part of the slack discussion I remember we had discussed documenting mitigation steps in case someone gets into the bad state of trying to delete the last system MachinePool and their MP is stuck in Deleting, can we file an issue to do this as part of a follow up if it's not done in this PR?

Sure will file an Issue for this and pick it up once this PR is merged.

**Comment by LochanRn** [comment]
@alexeldeib The Validator and Defaulter interface which I customized so that it can pass a client, I thought i can raise a PR to controller-runtime and figured out it is not promoted by the controller-runtime. It is because using a client in a webhook would make the api depend on specific controller-runtime version. Will this have any effect in capz code ?

https://github.com/kubernetes-sigs/controller-runtime/issues/1505#issuecomment-836099395

File: docs/book/src/topics/managedcluster.md [hunk 14]
```diff
@@ -214,3 +239,83 @@ Current limitations
 - Only supports Standard load balancer (SLB).
   - We will not support Basic load balancer in CAPZ. SLB is generally
     the path forward in Azure.
+
+## Troubleshooting
+
+1.) If a user tries to delete the machinepool which refers to the last system node pool AMMP webhook will reject deletion, so time stamp never gets set on the AMMP. However the timestamp would be set on the MP and would be in deletion state. To recover from this state create a new MP manually referencing the AMMP, fudge required references and finalizers in MP. In the AMMP remove the owner reference to the old MP, and set it to the new MP. Once the new MP is pointing to the AMMP you can delete the old MP. To delete the old MP remove the finalizers in that object.
```
Comment (alexeldeib) [inline docs/book/src/topics/managedcluster.md:245]: nit: as much as I like "fudge required references and finalizers", _slightly_ more professional phrasing might be nice 🙂 

perhaps: "manually edit the required reference and finalizers to link a new MachinePool to the AzureManagedMachinePool and unlink the old MachinePool"

File: docs/book/src/topics/managedcluster.md [hunk 15]
```diff
@@ -214,3 +239,83 @@ Current limitations
...
+  - apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: Cluster
+    name: capz-managed-aks
+    uid: 152ecf45-0a02-4635-987c-1ebb89055fa2
+  uid: ae4a235a-f0fa-4252-928a-0e3b4c61dbea
+spec:
+  clusterName: capz-managed-aks
+  minReadySeconds: 0
+  providerIDList:
+  - azure:///subscriptions/9107f2fb-e486-a434-a948-52e2929b6f18/resourceGroups/MC_rg_capz-managed-aks_eastus/providers/Microsoft.Compute/virtualMachineScaleSets/aks-agentpool0-10226072-vmss/virtualMachines/0
```
Comment (alexeldeib) [inline docs/book/src/topics/managedcluster.md:270]: awesome example, if this is a real susbcription ID please remove it (and ideally rebase)

File: exp/api/v1alpha4/azuremanagedmachinepool_webhook.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// log is for logging in this package.
+var azuremanagedmachinepoollog = logf.Log.WithName("azuremanagedmachinepool-resource")
```
Comment (alexeldeib) [inline exp/api/v1alpha4/azuremanagedmachinepool_webhook.go:35]: hrm...this follows an existing pattern, but I'm not convinced it's guaranteed to be defined (requires log.SetLogger(realLogger) to be called first, which we do in main.go, but I'd much prefer to pass this down via constructor, or as part of the functions somehow.

fine as-is, but making a note to my future self

File: exp/controllers/azuremangedcontrolplane_reconciler.go [hunk 17]
```diff
@@ -209,27 +210,63 @@ func (r *azureManagedControlPlaneReconciler) reconcileManagedCluster(ctx context
...
+
+			ammps = append(ammps, ammp)
+		}
+		if len(ammps) == 0 {
+			scope.Logger.Info("owner ref for system machine pools not ready")
+			return nil
+		}
 		// Add to cluster spec
-		managedClusterSpec.AgentPools = []managedclusters.PoolSpec{defaultPoolSpec}
+		managedClusterSpec.AgentPools = ammps
```
Comment (alexeldeib) [inline exp/controllers/azuremangedcontrolplane_reconciler.go:269]: we're sure we want all pools here, not just one system pool? fine with me, just wanted to check what you think.

File: util/webhook/validator.go (hunk 9 above)
Comment (alexeldeib) [reply util/webhook/validator.go:75]: yeah just to be clear this is totally fine, I suspect someone in CR didn't think too carefully about those few lines

**Comment by alexeldeib** [comment]
/lgtm 

thanks a bunch for all your work! my comments are all nits/confirmation, I'm good to merge this as-is 👍

File: docs/book/src/topics/managedcluster.md (hunk 15 above)
Comment (LochanRn) [reply docs/book/src/topics/managedcluster.md:270]: it is a fake subscription ID

File: docs/book/src/topics/managedcluster.md (hunk 14 above)
Comment (LochanRn) [reply docs/book/src/topics/managedcluster.md:245]: ack !

Comment (alexeldeib) [reply docs/book/src/topics/managedcluster.md:245]: great docs, by the way!

File: exp/controllers/azuremangedcontrolplane_reconciler.go (hunk 17 above)
Comment (LochanRn) [reply exp/controllers/azuremangedcontrolplane_reconciler.go:269]: I guess yes, as of now it seems fine, in future if this is creating any issue we can may be change it to just one.

Comment (alexeldeib) [reply exp/controllers/azuremangedcontrolplane_reconciler.go:269]: once we get e2e rock solid, we can a/b test the total e2e time with both approaches and decide from the data

Comment (LochanRn) [reply exp/controllers/azuremangedcontrolplane_reconciler.go:269]: yeah that makes sense 👍

**Comment by alexeldeib** [comment]
/lgtm 

proactively reapproving given the [force push diff](https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/833c2de0b68cd67d2e111d5b9993abcfdadf5275..81e3755302326c32e7ecc1a6bc3f813e65c6366c) (and hey I just figured out how to do that)

**Comment by CecileRobertMichon** [comment]
> proactively reapproving given the force push diff (and hey I just figured out how to do that)

TIL :D 

/approve
